src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,224 @@
+        /// </summary>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="IsAllocated"/> is <see langword="false"/>.</exception>
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            IntPtr handle = _handle;
+
+            if (handle == IntPtr.Zero)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+            }
+
+            object? target = nGetPrimaryAndSecondary(handle, out object? secondary);
+
+            return (target, secondary);
+        }
+
+        /// <summary>
+        /// Gets the target object instance for the current handle.
+        /// </summary>
+        /// <returns>The target object instance, if present.</returns>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal object? UnsafeGetTarget()
+        {
+            return nGetPrimary(_handle);
+        }
+
+        /// <summary>
+        /// Sets the target object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetTarget(object? target)
+        {
+            nSetPrimary(_handle, target);
+        }
+
+        /// <summary>
+        /// Sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Dependent"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetDependent(object? dependent)
+        {
+            nSetSecondary(_handle, dependent);
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <param name="dependent">The dependent instance, if available.</param>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <remarks>
+        /// This method mirrors <see cref="GetTargetAndDependent"/>, but without the allocation check.
+        /// The signature is also kept the same as the one for the internal call, to improve the codegen.
+        /// </remarks>
+        internal object? UnsafeGetTargetAndDependent(out object? dependent)
+        {
+            return nGetPrimaryAndSecondary(_handle, out dependent);
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        public void Dispose()
+        {
+            // Forces dependentHandle back to non-allocated state (if not already there)
+            // and frees the handle if needed.
+
+            if (_handle != IntPtr.Zero)
+            {
+                IntPtr handle = _handle;
+                _handle = IntPtr.Zero;
+                nFree(handle);
```

Comment:
I know there's been some discussion on this, but...

GCHandle.Free does use an interlocked.  Do we know why?  Are there uses of GCHandle that end up depending on that?  And if yes, do such uses not apply to DependentHandle as well?

I'm fine without the interlocked if we believe it's really unnecessary. (It's obviously there for a different use case than free'ing two copies of the same handle.)

Reply:
Not entirely sure why does `GCHandle` use an interlocked exchange, considering that it still has the same issue with respect to value copies anyway. We've been talking about this aspect too with @jkotas (see [comment above](https://github.com/dotnet/runtime/pull/54246#discussion_r652260942)), and also with @tannergooding on Discord. For instance, Tanner mentioned how `MemoryHandle` that was added recently also implements `IDisposable`, but has the same issue with respect to concurrent/copied frees, and so that it would probably be fine to just not use an interlocked here either. My argument for not using interlocked here was also that it could potentially be confusing for consumers given that the type itself would still not be thread-safe anyway (eg. calling `Dispose` together with any other API could still result in a double free), so I was thinking that just not making any individual method thread-safe and clearly documenting the whole type as not being thread-safe would be easier to understand for people using `DependentHandle` as well. Thoughts? 🙂

Reply:
I'd probably defer to @Maoni0/@jkotas on if it being interlocked is actually important.

It would still be impactful if it were the field of a class and was being used/passed byref.

Reply:
If it helps, Interlocked.CompareExchange was added to GCHandle.Free as part this change:
```
change 910279 edit on 2004/08/31 15:02:42

	New MDA DCR - detects when bad GCHandles are retrieved.
```
This changed introduced a parallel tracking structure for live GC handles that was automatically enabled when running under debugger. It was important for this parallel tracking structure to stay in sync with the actual GC handles and to be robust against incorrect GCHandle use (the point of the MDA was to catch incorrect uses of GCHandles). InterlockedCompare helped to achieve it.

The MDA is long gone, but the InterlockedCompareExchange stayed.

Reply:
If we keep the impl as is, we should explicitly say in the docs that the Target and Dependent properties are thread safe and that Dispose is not thread safe.

Reply:
Added some notes both on the type and on the individual APIs in 9c23a947ce5da21fb6169b6dae16a38f1412e8e9 🙂